kafka_producer.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright © 2024 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package kafka contains an implementation of Producer interface that can be
+// used to produce (that is send) messages to properly configured Kafka broker.
+package main
+
+// Generated documentation is available at:
+// https://pkg.go.dev/github.com/RedHatInsights/ccx-notification-service/producer
+//
+// Documentation in literate-programming-style is available at:
+// https://redhatinsights.github.io/ccx-notification-service/packages/producer/kafka/kafka_producer.html
+
+import (
+	"encoding/json"
+	"strings"
+	"time"
+
+	types "github.com/RedHatInsights/insights-results-types"
+	"github.com/Shopify/sarama"
+	"github.com/rs/zerolog/log"
+)
+
+const (
+	// StatusReceived is reported when a new payload is received.
+	StatusReceived = "received"
+	// StatusMessageProcessed is reported when the message of a payload has been processed.
+	StatusMessageProcessed = "processed"
+	// StatusSuccess is reported upon a successful handling of a payload.
+	StatusSuccess = "success"
+	// StatusError is reported when the handling of a payload fails for any reason.
+	StatusError = "error"
+)
+
+// Producer is a holder of a Sarama Producer and its configuration
+type Producer struct {
+	Configuration *BrokerConfiguration
+	Producer      sarama.SyncProducer
+}
+
+// PayloadTrackerProducer a holder of a Producer and the service name used
+// when sending tracking information
+type PayloadTrackerProducer struct {
+	ServiceName string
+	Producer
+}
+
+// PayloadTrackerMessage represents content of messages
+// sent to the Payload Tracker topic in Kafka.
+type PayloadTrackerMessage struct {
+	Service   string `json:"service"`
+	RequestID string `json:"request_id"`
+	Status    string `json:"status"`
+	Date      string `json:"date"`
+}
+
+// NewProducer instantiates a Kafka Producer
+func NewProducer(config *BrokerConfiguration) (*Producer, error) {
```

Comment:
do we not have something like this in other repos?

Reply:
we do, and I started the PR by refactoring and putting common code together in io-utils and ran into an issue with dependencies. So I decided to leave the refactor for another time, again 😞

Reply:
shall we create a task to unify producers in our Go code (?) this way this PR won't be blocked but we will eventually reduce code duplicity